src/components/List.tsx

```diff
@@ -0,0 +1,159 @@
+export const typedMemo: (<T>(c: T) => T) = memo;
+
+type OptionKey = string | number;
+
+interface BaseProps<D, P, K extends OptionKey> {
+    data: D[];
+    keySelector(datum: D, index: number): K;
+    renderer: React.ComponentType<P>;
+    rendererClassName?: string;
+    rendererParams: (key: K, datum: D, index: number, data: D[]) => P;
+}
+
+interface GroupOptions<D, P, K extends OptionKey, GP, GK extends OptionKey> {
+    groupComparator?: (a: GK, b: GK) => number;
+    groupKeySelector(datum: D): GK;
+    groupRenderer: React.ComponentType<GP>;
+    groupRendererClassName?: string;
+    groupRendererParams: (key: GK, index: number, data: D[]) => GP;
+    grouped: true;
+}
+
+interface NoGroupOptions {
+    grouped?: false;
+}
+
+export type ListProps<D, P, K extends OptionKey, GP, GK extends OptionKey> = (
+    BaseProps<D, P, K> & (GroupOptions<D, P, K, GP, GK> | NoGroupOptions)
+);
+
+export type GroupedListProps<D, P, K extends OptionKey, GP, GK extends OptionKey> = (
+    BaseProps<D, P, K> & GroupOptions<D, P, K, GP, GK>
+);
+
+function hasGroup<D, P, K extends OptionKey, GP, GK extends OptionKey>(
+    props: ListProps<D, P, K, GP, GK>,
+): props is (BaseProps<D, P, K> & GroupOptions<D, P, K, GP, GK>) {
+    return !!(props as BaseProps<D, P, K> & GroupOptions<D, P, K, GP, GK>).grouped;
+}
+
+function GroupedList<D, P, K extends OptionKey, GP, GK extends OptionKey>(
+    props: GroupedListProps<D, P, K, GP, GK>,
+) {
+    const {
+        groupKeySelector,
+        groupComparator,
+        renderer: Renderer,
+        groupRenderer: GroupRenderer,
+        groupRendererClassName,
+        groupRendererParams,
+        data,
+        keySelector,
+        rendererParams,
+        rendererClassName,
+    } = props;
+
+    const renderListItem = (datum: D, i: number) => {
+        const key = keySelector(datum, i);
+        const extraProps = rendererParams(key, datum, i, data);
+
+        return (
+            <Renderer
+                key={key}
+                className={rendererClassName}
+                {...extraProps}
+            />
+        );
+    };
+    const renderGroup = (groupKey: GK, index: number) => {
+        const extraProps = groupRendererParams(groupKey, index, data);
+
+        return (
+            <GroupRenderer
+                key={groupKey}
+                className={groupRendererClassName}
+                {...extraProps}
+            />
+        );
+    };
+
+    const groups = useMemo(
+        () => listToGroupList(data, groupKeySelector),
+        [data, groupKeySelector],
+    );
+
+    const sortedGroupKeys = useMemo(
+        () => {
+            const keys = Object.keys(groups) as GK[];
+            return keys.sort(groupComparator);
+        },
+        [groups, groupComparator],
+    );
+
+    const children: React.ReactNode[] = [];
+    sortedGroupKeys.forEach((groupKey, i) => {
+        children.push(renderGroup(groupKey, i));
+        children.push(...groups[groupKey].map(renderListItem));
```

Comment:
While creating children inside grouped list, the key should change when group changes